subprocess_utils.hpp

```diff
@@ -0,0 +1,219 @@
+#pragma once
+
+#include <string>
+#include <iostream>
+
+// unix process stuff
+#include <signal.h>
+#include <unistd.h>
+#include <sys/wait.h>
+#include <sys/prctl.h>
+#include <cstring>
+
+
+namespace subprocess {
+    
+    /**
+     * A TwoWayPipe that allows reading and writing between two processes
+     * must call initialize before being passed between processes or used
+     * */
+class TwoWayPipe {
+	private:
+		//[0] is the output end of each pipe and [1] is the input end of each pipe
+		int input_pipe_file_descriptor[2];
+		int output_pipe_file_descriptor[2];
+        std::string internalBuffer;
+        bool inStreamGood = true;
+        
+	public:
+		TwoWayPipe() = default;
+		
+		/**
+		 * initializes the TwoWayPipe the pipe can not be used until this is called
+		 * */
+		void initialize(){
+			pipe(input_pipe_file_descriptor);
+			pipe(output_pipe_file_descriptor);
+		}
+	private:
+	    /**
+	     * closes the ends that aren't used (do we need to do this?
+	     * */
+		void closeUnusedEnds() {
+			//we don't need the input end of the input pipe or the output end of the output pipe
+			close(input_pipe_file_descriptor[1]);
```

Comment:
is this the case for both children and parent? you'll need to double check this - but if this is intended to only be called for the child, we may want to use inheritance to convey this (or really lazily, a bool that signifies whether it's a child or not)

but the class could use a little more jiggling around, it's currently pretty easy to mess up with it (calling setAsChildEnd twice), where this is something that makes sense to be done at most once), and some functions are mutually exclusive, but they don't enforce them internally

Reply:
I agree, hard to do it with inheritance since it needs to be created before being split between processes but a flag to prevent calling it twice would work. It should only be called internally from the process class anyway so maybe I'll just make the pipe inaccessible and wrap io calls in public Process calls.